state/initialize.go

```diff
@@ -0,0 +1,368 @@
+				logger.Errorf("error closing state while aborting Initialize: %v", closeErr)
+			}
+		}
+	}()
+	st.controllerModelTag = modelTag
+
+	// A valid model is used as a signal that the
+	// state has already been initalized. If this is the case
+	// do nothing.
+	if _, err := st.Model(); err == nil {
+		return nil, errors.New("already initialized")
+	} else if !errors.IsNotFound(err) {
+		return nil, errors.Trace(err)
+	}
+
+	logger.Infof("initializing controller model %s", modelTag.Id())
+
+	modelOps, modelStatusDoc, err := st.modelSetupOps(
+		args.ControllerConfig.ControllerUUID(),
+		args.ControllerModelArgs,
+		&lineage{
+			ControllerConfig: args.ControllerInheritedConfig,
+			RegionConfig:     args.RegionInheritedConfig,
+		})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	salt, err := utils.RandomSalt()
+	if err != nil {
+		return nil, err
+	}
+
+	dateCreated := st.NowToTheSecond()
+	ops := createInitialUserOps(
+		args.ControllerConfig.ControllerUUID(),
+		args.ControllerModelArgs.Owner,
+		args.MongoInfo.Password,
+		salt,
+		dateCreated,
+	)
+	ops = append(ops,
+
+		txn.Op{
+			C:      controllersC,
+			Id:     modelGlobalKey,
+			Assert: txn.DocMissing,
+			Insert: &controllersDoc{
+				CloudName: args.Cloud.Name,
+				ModelUUID: st.ModelUUID(),
+			},
+		},
+		createCloudOp(args.Cloud),
+		txn.Op{
+			C:      controllersC,
+			Id:     apiHostPortsKey,
+			Assert: txn.DocMissing,
+			Insert: &apiHostPortsDoc{},
+		},
+		txn.Op{
+			C:      controllersC,
+			Id:     stateServingInfoKey,
+			Assert: txn.DocMissing,
+			Insert: &StateServingInfo{},
+		},
+		txn.Op{
+			C:      controllersC,
+			Id:     hostedModelCountKey,
+			Assert: txn.DocMissing,
+			Insert: &hostedModelCountDoc{},
+		},
+		createSettingsOp(controllersC, controllerSettingsGlobalKey, args.ControllerConfig),
+		createSettingsOp(globalSettingsC, controllerInheritedSettingsGlobalKey, args.ControllerInheritedConfig),
+	)
+	for k, v := range args.Cloud.RegionConfig {
+		// Create an entry keyed on cloudname#<key>, value for each region in
+		// region-config. The values here are themselves
+		// map[string]interface{}.
+		ops = append(ops, createSettingsOp(globalSettingsC, regionSettingsGlobalKey(args.Cloud.Name, k), v))
+	}
+
+	for tag, cred := range args.CloudCredentials {
+		ops = append(ops, createCloudCredentialOp(tag, cred))
+	}
+	ops = append(ops, modelOps...)
+
+	if err := st.runTransaction(ops); err != nil {
+		return nil, errors.Trace(err)
+	}
+	controllerTag := names.NewControllerTag(args.ControllerConfig.ControllerUUID())
+	if err := st.start(controllerTag); err != nil {
+		return nil, errors.Trace(err)
+	}
+	probablyUpdateStatusHistory(st, modelGlobalKey, modelStatusDoc)
+	return st, nil
+}
+
+// InitDatabase creates all the collections and indices in a Juju database.
+func InitDatabase(session *mgo.Session, modelUUID string, settings *controller.Config) error {
+	schema := allCollections()
+	err := schema.Create(
```

Comment:
Existing code, however wrapping seems completely unnecessary here, in fact it could be nested:

if err := schema.Create(session.DB(jujuDB), settings); err != nil {